File: Data Structures and Algorithms/C++ Version/BinarySearch.cpp
#include <iostream>
#include <vector>
#include<tuple>

std::tuple<int, int> binarySearch (std::vector<int> items, int target) {
    int steps = 0;
    int idx = 0;

    while (true) {
        int mid = items.size() / 2;

        if (target == items[mid]) {
            steps += 1;
            idx += mid;
            std::tuple<int, int> t = {steps, idx};
            return t;
        }

        if (mid == 0) {
            std::tuple<int, int> t = {0, 0};
            return t;
        }

        else if (target > items[mid]) {
            steps += 1;
            items = std::vector<int>(items.begin() + mid + 1, items.end());
        }

        else if (target < items[mid]) {
            steps += 1;
            items = std::vector<int>(items.begin(), items.end() - mid - 1);
        }
    }

}

int main () {
    std::vector<int> items = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    int target = 9;

    std::tuple<int, int> result;
    int steps;
    int idx;

    result = binarySearch(items, target);
    steps = std::get<0>(result);
    idx = std::get<1>(result);

    std::cout << "Steps: " << steps << ", " << "Index of element: " << idx;
    return 0;
}

File: Data Structures and Algorithms/Grokking Algorithms (Book by Aditya Y.Bhargava)/BinarySearch.py


## Items must be a sorted list.
def binarySearch(items: list, target):
    """
    Perform binary search on a sorted list to find the target element.

    Args:
        items (list): A sorted list of comparable elements.
        target: The element to search for in the list.

    Returns:
        tuple: A tuple containing the number of steps taken to find the target
               and the index of the target in the list. If the target is not found,
               returns (None, None).
    """
    step = 0
    while True:
        mid = len(items) // 2
        if target == items[mid]:
            step += 1
            idx = mid
            return step, idx
        if mid == 0:
            return None, None
        elif target > items[mid]:
            step += 1
            items = items[mid+1:]
        elif target < items[mid]:
            items = items[:mid]
            step += 1

#   For any list of n, binary search will take log2n steps to run in the worst case.
items = [1, 2, 3, 4, 5, 6, 7, 8, 9]
target = 9
steps, idx = binarySearch(items, target)
print(f"Steps: {steps}, Index of element: {idx}")